Add method to allow setting multiple attributes at once to … (#1473)
open-telemetry/opentelemetry-specification: specification/trace/api.md

  as arguments. This MAY be called `SetAttribute`. To avoid extra allocations some
  implementations may offer a separate API for each of the possible value types.

The Span interface MAY provide:

> - An API to set multiple `Attributes` at once, where all `Attributes` are passed as a
>   single argument. This MAY be called `SetAllAttributes`.

Reviewer: ```suggestion
- An API to set multiple `Attributes` at once, where the `Attributes` are passed as a
  in a single method call. This MAY be called `SetAllAttributes`.
```

Whether this would be a single (map) argument or multiple arguments (e.g. `**kwargs` in Python) is language dependent.

Author: Yup thanks

>   single method call. This method MAY be called `SetAllAttributes`.

Reviewer 2: I would vote for `SetAttributes` instead of `SetAllAttributes`.
`SetAllAttributes` gives a feeling that it will make previous calls invalid:
```text
span.SetAllAttributes(foo: 1, bar: 2);
span.SetAllAttributes(baz: 3);
// do we have foo, bar and baz, or we only have baz?
```

```text
span.SetAttributes(foo: 1, bar: 2);
span.SetAttributes(baz: 3);
// we know that we will have foo, bar and baz
```

Author: Heh this is the obvious of Java, we would expect `SetAttributes` to clear existing ones and `SetAll` to keep them, similar to `putAll` in a Map. I can remove the naming recommendation if it seems too language specific?

Reviewer 3: @anuraaga Probably yes. We had in the past a (now forgotten) PR that tried to add this, and it was suggested to define the operation as a very general thing, and don't make the name a hard requirement (and mention languages can adapt it to their own guidelines, etc).

Reviewer 4: `AddAttributes` is shorter and unambiguous

Author: Sorry for the delay, went ahead and removed the naming recommendation.

> AddAttributes is shorter and unambiguous

I think it's a bit ambiguous, previous attributes (especially for list attributes) we could expect to remain in some form if it's just `Add`.
